<meta charset="utf-8"/>
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    // Don't process any delimiters, only <script type="math/tex">
    // tags generated by kramdown from $$...$$ in source.
    // (Could also avoid loading tex preprocessor - only need
    // tex input jax - but not worth the trouble.)
    tex2jax: {
      inlineMath: [],
      displayMath: [],
    }
  });
</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>

## Course description
Programming skills and software tools for building automated bioinformatics pipelines and computational biology analyses. Emphasis on UNIX tools and R libraries for distilling raw sequencing data into interpretable results.  This course is aimed at students familiar with UNIX and with some programming experience in python, R, or C/C++.

## Instructional staff
* [Cole Trapnell](http://cole-trapnell-lab.github.io/)
* Teaching assistant: April Lo

Please click on the links above for email addresses and office locations.

## Meeting times and locations
### Classes: 

Monday and Wednesday, 10:30 - 12:20 am, [zoom](http://www.washington.edu/home/maps/southcentral.html?gnom).

### Class Slack: 
We will use Slack during class and outside of class to communicate, share code snippets, ask and answer questions. The class slack is here:
* [gs-bioinf-workflows.slack.com](https://gs-bioinf-workflows.slack.com)

You will receive an invitation to join prior to the first class.

### Office hours: 
* No official office hours. Post questions on Slack as needed.

## Prerequisites
* Substantial background in molecular and cellular biology, genetics, biochemistry, or related disciplines.
* Familiarity with UNIX.
* Some programming experience in python, R, or C/C++. 
* Students are encouraged to have taken GENOME559 and/or GENOME560.

## Course requirements
* The course involves hands-on programming during class time. We will use the GS compute cluster, so make sure you can log into it from your computer remotely.
* All programming projects are due by the start of class on the date listed.
* You are welcome to talk to classmates about principles for solving problems, but please do not share code or program together. In many ways, writing your own code is where you will learn the most for this class.

## Examinations

There will be no examinations.

## Course grade
Grades will come 50% from the programming projects and 50% from class participation. However, given the extraordinary circumstances surrounding the COVID-19 outbreak, the grading rubric will be revised as needed.  


## Course materials
We will read from several online resources and tutorials. I strongly encourage you to read all of the material in the following:

* [Comprehensive single-cell transcriptional profiling of a multicellular organism (Packer et al)](http://cole-trapnell-lab.github.io/papers/cao-scrnaseq-combinatorial-indexing/)
* [Git Basics](https://www.freecodecamp.org/news/learn-the-basics-of-git-in-under-10-minutes-da548267cc91/)
* [Pro Git](https://git-scm.com/book/en/v2)
* [BASH basics](https://tldp.org/HOWTO/Bash-Prog-Intro-HOWTO.html)
* [Essential UNIX](https://www.unixtutorial.org/basic-unix-commands)
* [Sed and Awk](https://learning.oreilly.com/library/view/sed-awk/1565922255/)
* [Sed and Awk, pocket ref](https://learning.oreilly.com/library/view/sed-and-awk/0596003528/)
* [STAR Manual](https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf)
* [SAM format](https://samtools.github.io/hts-specs/SAMv1.pdf)
* [samtools](http://www.htslib.org/doc/samtools.html)
* [BED format](http://genome.ucsc.edu/FAQ/FAQformat#format1)
* [bedtools](https://bedtools.readthedocs.io/en/latest/)
* [R Markdown: the definitive guide](https://bookdown.org/yihui/rmarkdown/)
* [R for Data Science](https://r4ds.had.co.nz/)

Specific, selected readings for the course will be listed in the course schedule below.

## Helpful software

* [Visual studio code](https://code.visualstudio.com/) - An outstanding code editor and integrated development environment
* [Rstudio](https://rstudio.com/) - An integrated development environment for R

## Class schedule

| Date |  Topic | Reading | 
| ---- | ------------- | ----------------- | 
|  3/30   | [Course overview, student setup, and version control](slides/class_1_intro/) | [Git Basics](https://www.freecodecamp.org/news/learn-the-basics-of-git-in-under-10-minutes-da548267cc91/; [Packer et al](http://cole-trapnell-lab.github.io/papers/cao-scrnaseq-combinatorial-indexing/))  | 
|  4/1  | [Intro to bioinformatics pipelines, automation](slides/class_2_automation) | [Essential UNIX](https://www.unixtutorial.org/basic-unix-commands); [BASH basics (sections 1-7)](https://tldp.org/HOWTO/Bash-Prog-Intro-HOWTO.html) | 
|   4/6   | [Tools for working with tables](slides/class_3_tools_for_tables) | [Sed and Awk](https://learning.oreilly.com/library/view/sed-awk/1565922255/) | 
|   4/8  | [NGS read alignment](slides/class_4_ngs_read_alignment)  | [SAM format](https://samtools.github.io/hts-specs/SAMv1.pdf); [bedtools](https://bedtools.readthedocs.io/en/latest/) | 
|   4/13  | [Electronic lab notebooks](slides/class_5_electronic_lab_notebooks)  | [R Markdown (chapter 3)](https://samtools.github.io/hts-specs/SAMv1.pdf); | 
|   4/15   | [](slides/XXX) | | 
|   4/20   | [](slides/XXX) | | 
|   4/22   | [](slides/XXX) |  
|   4/27  |  | |
|   4/29  |  | |

## Example files

* [qsub job script](example_files/example_qsub.sh)
* [grep & sed example 1](example_files/grep_sed_example1.txt)
* [grep & sed example 2](example_files/grep_sed_example2.txt)
* [jedi table 1](example_files/jedi_example1.txt)
* [jedi table 2](example_files/jedi_example2.txt)
* [jedi table 3](example_files/jedi_example3.txt)
* [sci-RNA-seq read1](example_files/exampleR1.fastq)
* [sci-RNA-seq read2](example_files/exampleR2.fastq)
* [small_reads.fastq](example_files/small_reads.fastq)
* [genes.bed](example_files/genes.bed)
